cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -3072,8 +3078,11 @@ Status S3FileSystem::Move(const std::string& src, const std::string& dest) {
   }
   RETURN_NOT_OK(impl_->CopyObject(src_path, dest_path));
   RETURN_NOT_OK(impl_->DeleteObject(src_path.bucket, src_path.key));
-  // Source parent may be implicitly deleted if it became empty, recreate it
-  return impl_->EnsureParentExists(src_path);
+  if (options().create_missing_dirs_on_delete) {
+    // Source parent may be implicitly deleted if it became empty, recreate it
+    return impl_->EnsureParentExists(src_path);
+  }
```

Comment:
For symmetry, when checking if a directory exists, don't consider objects with a `dir/` prefix in the path other than the empty directory marker itself as a sign that the directory exists.

Reply:
Is my understanding correct?

Currently, when attempting to delete a directory, it does not check for the presence of other files within the directory but instead creates a marker.

Are you suggesting these improvements?

- When a specific directory is being deleted, it is necessary to check for the existence of the parent directory:
  - If there are no files in the parent directory and no marker exists, then create a marker.
  - If there are files in the parent directory, do not create a marker.
- When a specific file is being deleted, it is necessary to check for the presence of other files within the directory:
  - If other files exist, do not create a marker.
  - If no other files exist, create a marker.

Reply:
Other parts of the filesystem implementation have an operation that "checks if a directory exists". That operation usually considers that keys with a prefix imply the existence of the directory even if a marker doesn't exist.

For instance, if the object store has only this object:

`dir0/myfile.parquet`

`DirExists?("dir0")` returns `true` because there is at least one object key with `dir0` as the prefix.

The problem now is that `DeleteFile("dir0/myfile.parquet")` (w/ this new option == true) will cause `DirExists?("dir0")` to return `false` after the last file is deleted. Which breaks the expectation of a filesystem.

But if you modify `DirExists?("dir0")` to return `true` only if a marker exists (w/ this new option == true) it will increase the chances that a marker is always created *before* files are added to the directory.

`CreateDir("dir0")` creates the `dir0/` marker. Creating `"dir0/myfile.parquet"` checks the directory exists (by checking the existence of the marker). Taking the object store to this list of keys:

```
dir0/
dir0/myfile.parquet
```

Now it's totally safe to delete `myfile.parquet` without caring about `dir0`.

Reply:
This means that if I start building my filesystem with this option set to true and no other way of adding objects to the store, empty directory markers will always be created (but only once), and deleting files without re-creating the empty directory marker is safe.

If you think that checking that directory marker exists is also too expensive on the other operations, there is already an option (the one above yours) controlling that. If both are set to true, then we really get that the user doesn't care about filesystem semantics at their own peril.